mantle/util/retry.go

```diff
@@ -72,6 +72,32 @@ func RetryUntilTimeout(timeout, delay time.Duration, f func() error) error {
 	return nil
 }
 
+// Like RetryUntilTimeout, however we will only execute f retries number of times.
+// We will also only execute f async, so that we may interrupt execution when needed.
```

Comment:
Let's go back to the original problem and work from there. We have an API call that didn't return for 30 minutes. We don't want to wait that long to find out something is not working. How do we handle this case?

Reply:
So to clarify the problem, my understanding is that when the main goroutine exits, the orphaned routines will also exit but in a non-graceful manner. Code that would clean up resources (like defers) will not be run. Its not ideal but perhaps we could shift the responsibility to the programmers to not use `f` such that we require resource cleanups? Not sure ...

I will improve the wording to reflect what is actually happening. However, as Dusty mentioned, we have run into this issue before and we have not been able to find a solution that doesn't leave goroutines running.

Reply:
@saqibali-2k The failure case is a concern, but less so.  We already have to handle non-graceful exit (kola can always dump core or OOM), which is why we have a separate mechanism for GCing leaked cloud resources.  The larger problem is that multiple tries might succeed, perhaps even after the function returns.  We shouldn't have an API that claims to do something straightforward (retry requests) but actually leaves a mess behind.

Since we don't need retries for our actual use case, let's drop that part of the code and focus on providing a timeout wrapper.  And let's document that it should only be used as a last resort, and maybe put `Unsafe` or something in the function name so it doesn't get used casually.

I assume you've already checked that the Azure Go API doesn't provide a timeout mechanism?

Reply:
> I assume you've already checked that the Azure Go API doesn't provide a timeout mechanism?

Maybe that's the answer!

https://github.com/coreos/coreos-assembler/blob/de628bc6d86fcea34b89b3cb3f019a78b0f40384/mantle/vendor/github.com/Azure/azure-sdk-for-go/arm/compute/virtualmachines.go#L192-L195